### PR TITLE
feat: Allow SLURM executor option `--mem-per-cpu`

### PR DESCRIPTION
**Pull request by mobilegenome**
This PR addresses #2054. 

Some SLURM cluster require memory to be defined per CPU using the `--mem-per-cpu` argument in the `sbatch` command and disallow using `--mem`. 

Currently, nextflow translates the `task.memory` directive into `--mem` only. 

The proposed changes introduces the boolean `executor.memPerCpu` option to execute `sbatch` with `--mem-per-cpu`. 

This is my first time working with the Nextflow codebase and using Groovy, so any help getting this to work is appreciated 🙏

### CODE REVIEW

**Comment by bentsherman** [comment]
@mobilegenome thanks for your contribution. Can you pull my edits and make sure that it works with a pipeline on your cluster?

**Comment by mobilegenome** [comment]
Thanks for fixing the code and making it neat! I was hopeful it works after seeing that the CI passes (didn't test my code before) but it runs into a NullPointerException. Seemingly, getting the QueueStatus broke, but I can't tell why. 

On the good side, the `sbatch` directives are correctly set in the `.command.run` file!

I tried to run `nextflow-io/hello`. This is the error message

```
Jun-15 09:15:56.012 [Task monitor] DEBUG nextflow.processor.TaskProcessor - Handling unexpected condition for
  task: name=sayHello (1); work-dir=/cluster/scratch/flamme/nextflow-test/work/a1/a91855b51c00a18963f9e0c8a5e4d0
  error [java.lang.NullPointerException]: java.lang.NullPointerException
Jun-15 09:15:56.028 [Task monitor] ERROR nextflow.processor.TaskProcessor - Error executing process > 'sayHello (1)'

Caused by:
  java.lang.NullPointerException

java.lang.NullPointerException: null
        at nextflow.util.Throttle.cache(Throttle.groovy:208)
        at nextflow.executor.AbstractGridExecutor.getQueueStatus(AbstractGridExecutor.groovy:311)
        at nextflow.executor.AbstractGridExecutor.checkStartedStatus(AbstractGridExecutor.groovy:357)
        at nextflow.executor.GridTaskHandler.isStarted(GridTaskHandler.groovy:430)
        at nextflow.executor.GridTaskHandler.checkIfRunning(GridTaskHandler.groovy:411)
        at nextflow.processor.TaskPollingMonitor.checkTaskStatus(TaskPollingMonitor.groovy:605)
        at nextflow.processor.TaskPollingMonitor.checkAllTasks(TaskPollingMonitor.groovy:534)
        at nextflow.processor.TaskPollingMonitor.pollLoop(TaskPollingMonitor.groovy:409)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:564)
        at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:107)
        at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323)
        at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1258)
        at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1035)
        at org.codehaus.groovy.runtime.InvokerHelper.invokePogoMethod(InvokerHelper.java:1036)
        at org.codehaus.groovy.runtime.InvokerHelper.invokeMethod(InvokerHelper.java:1019)
        at org.codehaus.groovy.runtime.InvokerHelper.invokeMethodSafe(InvokerHelper.java:97)
        at nextflow.processor.TaskPollingMonitor$_start_closure2.doCall(TaskPollingMonitor.groovy:293)
        at nextflow.processor.TaskPollingMonitor$_start_closure2.call(TaskPollingMonitor.groovy)
        at groovy.lang.Closure.run(Closure.java:493)
        at java.base/java.lang.Thread.run(Thread.java:832)
Jun-15 09:15:56.041 [Task monitor] DEBUG nextflow.Session - Session aborted -- Cause: java.lang.NullPointerException
```

**Comment by bentsherman** [comment]
This error seems unrelated. Do you have any custom config settings for your slurm executor? It looks like the `queueInterval` is null.

**Comment by bentsherman** [comment]
Nevermind, I found a bug in my changes. Fixed and tested in our SLURM environment, so let's move it forward!

**Review by bentsherman — APPROVED** [review]

**Comment by mobilegenome** [comment]
Ace, thanks so much for working on this! Will test it on my side as well, but I have full confidence it works now.

**Comment by mobilegenome** [comment]
@fa2k that is actually an excellent idea, I didn't think of it! Would you mind adding this functionality and corresponding documentation? I'm about to leave to for a week of holidays and would only have time doing this in a week. Happy to do it myself then, but if you want to keep this moving in the meantime, feel free to do so.

**Comment by bentsherman** [comment]
@fa2k That sounds good to me, I will add it

File: docs/config.md [hunk 1]
```diff
@@ -605,6 +605,12 @@ The following settings are available:
 `executor.memory`
 : The maximum amount of memory made available by the underlying system. Used only by the `local` executor.
 
+`executor.memPerCpu`
```
Comment (pditommaso) [inline docs/config.md:608]: Considering already exist `perJobMemLimit` and `perTaskReserve`, it'd preferable to call this setting `perCpuMemAllocation` for consistency

**Comment by pditommaso** [comment]
Also, a note it should be added in the SLURM docs explaining what this setting does as done for LSF for `perJobMemLimit`. See [here](https://www.nextflow.io/docs/edge/executor.html#lsf-executor)

**Comment by pditommaso** [comment]
Nice, thanks

**Pull request merged**